std/datetime/stopwatch.d

```diff
@@ -0,0 +1,425 @@
+          $(REF TickDuration,core,time) are in std.datetime.package and would
+          conflict with the symbols in this module. After the old symbols have
+          gone through the deprecation cycle and have been removed, then this
+          module will be publicly imported in std.datetime.package.)
+
+    License:   $(HTTP www.boost.org/LICENSE_1_0.txt, Boost License 1.0).
+    Authors:   Jonathan M Davis and Kato Shoichi
+    Source:    $(PHOBOSSRC std/datetime/_stopwatch.d)
++/
+module std.datetime.stopwatch;
+
+public import core.time;
+import std.typecons : Flag;
+
+/++
+   Used by StopWatch to indicate whether it should start immediately upon
+   construction.
+
+   If set to $(D AutoStart.no), then the StopWatch is not started when it is
+   constructed.
+
+   Otherwise, if set to $(D AutoStart.yes), then the StopWatch is started when
+   it is constructed.
+  +/
+alias AutoStart = Flag!"autoStart";
+
+
+/++
+   StopWatch is used to measure time just like one would do with a physical
+   stopwatch, including stopping, restarting, and/or resetting it.
+
+   $(REF MonoTime,core,time) is used to hold the time, and it uses the system's
+   monotonic clock, which is high precision and never counts backwards (unlike
+   the wall clock time, which $(I can) count backwards, which is why
+   $(REF SysTime,std,datetime,systime) should not be used for timing).
+
+   Note that the precision of StopWatch differs from system to system. It is
+   impossible for it to be the same for all systems, since the precision of the
+   system clock and other system-dependent and situation-dependent factors
+   (such as the overhead of a context switch between threads) varies from system
+   to system and can affect StopWatch's accuracy.
+  +/
+struct StopWatch
+{
+public:
+
+    ///
+    @system nothrow @nogc unittest
+    {
+        import core.thread : Thread;
+
+        auto sw = StopWatch(AutoStart.yes);
+
+        Duration t1 = sw.peek();
+        Thread.sleep(usecs(1));
+        Duration t2 = sw.peek();
+        assert(t2 > t1);
+
+        Thread.sleep(usecs(1));
+        sw.stop();
+
+        Duration t3 = sw.peek();
+        assert(t3 > t2);
+        Duration t4 = sw.peek();
+        assert(t3 == t4);
+
+        sw.start();
+        Thread.sleep(usecs(1));
+
+        Duration t5 = sw.peek();
+        assert(t5 > t4);
+
+        // If stopping or resetting the StopWatch is not required, then
+        // MonoTime can easily be used by itself without StopWatch.
+        auto before = MonoTime.currTime;
+        // do stuff...
+        auto timeElapsed = MonoTime.currTime - before;
+    }
+
+    /++
+        Constructs a StopWatch. Whether it starts immediately depends on the
+        $(LREF AutoStart) argument.
+
+        If $(D StopWatch.init) is used, then the constructed StopWatch isn't
+        running (and can't be, since no constructor ran).
+      +/
+    this(AutoStart autostart) @safe nothrow @nogc
```

Comment:
Aside from whether that would help or harm usability, it wouldn't work in D, because then it would be a default constructor.